tower-balance/src/lib.rs

```diff
@@ -26,166 +53,158 @@ pub enum Error<T, U> {
+    D,
+    load::WithConstant<D::Service>,
+    select::RoundRobin<D::Key, load::Constant<D::Service>>,
+>
+where
+    D: Discover,
+    D::Key: Clone,
+{
+    /// Creates a new balancer.
     ///
-    /// This function may panic if `poll_ready` did not return `Ready`
-    /// immediately prior.
-    fn next(&mut self) -> &mut T::Service {
-        let pos = self.pos;
-        self.inc_pos();
-
-        match self.endpoints.get_index_mut(pos) {
-            Some((_, val)) => val,
-            _ => panic!(),
-        }
+    /// The balancer selects each node in order. This order is not strictly enforced.
+    pub fn round_robin(discover: D) -> Self {
+        let load = load::WithConstant::min();
+        let select = select::RoundRobin::default();
+        Self::new(discover, load, select)
     }
+}
 
-    // ===== internal =====
+impl<D, R> Balance<
+    D,
+    load::WithPendingRequests<D::Service>,
+    select::PowerOfTwoChoices<D::Key, load::PendingRequests<D::Service>, R>,
+>
+where
+    D: Discover,
+    D::Key: Clone,
+    R: Rng,
+{
+    /// Creates a new balancer.
+    ///
+    /// The balancer selects the least-loaded endpoint using Mitzenmacher's _Power of Two
+    /// Choices_.
+    pub fn pending_requests_p2c(discover: D, rng: R) -> Self {
+        let load = load::WithPendingRequests::default();
+        let select = select::PowerOfTwoChoices::new(rng);
+        Self::new(discover, load, select)
+    }
+}
 
-    fn update_endpoints(&mut self) -> Result<(), T::DiscoverError> {
-        use tower_discover::Change::*;
+impl<D, L, S> Service for Balance<D, L, S>
+where
+    D: Discover,
+    D::Key: Clone,
+    L: WithLoad<Service = D::Service>,
+    S: Select<Key = D::Key, Service = L::PollLoad>,
+{
+    type Request = <L::PollLoad as Service>::Request;
+    type Response = <L::PollLoad as Service>::Response;
+    type Error = Error<<L::PollLoad as Service>::Error, D::DiscoverError>;
+    type Future = ResponseFuture<<L::PollLoad as Service>::Future, D::DiscoverError>;
 
-        loop {
-            let change = match try!(self.discover.poll()) {
-                Async::Ready(change) => change,
-                Async::NotReady => return Ok(()),
-            };
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        try_ready!(self.poll_discover());
+        assert!(!self.endpoints.is_empty(), "Balance is ready when there are endpoints");
 
-            match change {
-                Insert(key, val) => {
-                    // TODO: What to do if there already is an entry with the
-                    // given `key` in the set?
-                    self.endpoints.entry(key).or_insert(val);
-                }
-                Remove(key) => {
-                    // TODO: What to do if there is no entry?
-                    self.endpoints.remove(&key);
-                }
-            }
+        let nodes = self.endpoints.iter_mut();
+        let key =  try_ready!(self.select.poll_next_ready(nodes).map_err(Error::Inner));
```

Comment:
I wonder if this could be called from `call`.

For this to be true, `poll_next_ready` would actually not be able to be async. It would have to always return a service.

Would it be sane for the load balancer to say that it is ready if at least one endpoint is ready? In this case, `poll_ready` would only need to iterate the list of not-ready endpoints, polling them. If any are ready, then the load balancer is ready itself. This would also simplify the implementation of `Select`.

Reply:
For this to be sync, we'd have to know the readiness of all nodes. I don't think we should expect the balancer to manage this totally (with 10^2-10^3 nodes).  Rather, the node selection strategy manages this polling so it only polls the nodes for which it actually needs state.